patchers/value/value.go

```diff
@@ -0,0 +1,260 @@
+// Package value provides a Patcher that uses interfaces to allow custom types that can be represented as standard go values to be used more easily in expressions.
+//
+// # Example Usage
+//
+//	import (
+//		"fmt"
+//		"github.com/antonmedv/expr/patchers/value"
+//		"github.com/antonmedv/expr"
+//	)
+//
+//	type customInt struct {
+//	       	Int int
+//	}
+//
+//	// Provides type checking at compile time
+//	func (v *customInt) IntValue() int {
+//	       	return v.Int
+//	}
+//
+//	// Lets us return nil if we need to
+//	func (v *customInt) ExprValue() any {
+//	       	return v.Int
+//	}
+//
+//	func main() {
+//		env := make(map[string]any)
+//		env["ValueOne"] = &customInt{1}
+//		env["ValueTwo"] = &customInt{2}
+//
+//		program, err := expr.Compile("ValueOne + ValueTwo", expr.Env(env), value.Patcher)
+//
+//		if err != nil {
+//			panic(err)
+//		}
+//
+//		out, err := vm.Run(program, env)
+//
+//		if err != nil {
+//			panic(err)
+//		}
+//
+//		fmt.Printf("Got %v", out)
+//	}
+package value
+
+import (
+	"reflect"
+	"time"
+
+	"github.com/antonmedv/expr"
+	"github.com/antonmedv/expr/ast"
+	"github.com/antonmedv/expr/conf"
+)
+
+// Patcher is an expr.Option that both patches the program and adds the `getExprValue` function.
+// Use it directly as an Option to expr.Compile()
+var Patcher = func() expr.Option {
+	vPatcher := patcher{}
+	return func(c *conf.Config) {
+		c.Visitors = append(c.Visitors, vPatcher)
+		vPatcher.ApplyOptions(c)
+	}
+}()
+
+// A ExprValuer provides a generic function for a custom type to return standard go values.
+// It allows for returning a `nil` value but does not provide any type checking at expression compile.
+//
+// A custom type may implement both ExprValuer and a type specific interface to enable both
+// compile time checking and the ability to return a `nil` value.
+type ExprValuer interface {
+	ExprValue() any
+}
+
+type IntValuer interface {
+	IntValue() int
```

Comment:
I'm not happy with the `Value` suffix either. The big issue is running into collisions.

Example of unwanted collisions:
`big.Float` - has `Int`, `Int64`, `Float32`, `Float64`, `Uint64`, and `String` as methods already. Signatures are mostly different, but I think it is reasonable someone might want to embed it and add and interface for `expr` use.  Complexity goes way up if you are all of a sudden dealing with naming collisions.

`fmt.Stringer` - `String` method. To start `time.Time` implements this, so special casing would be required or we would turn all times into strings. I also don't think it's unreasonable for a custom type to implement `String` for printing, but want to use a different interface with `expr`. Multiple matching interfaces leads to the question of precedence, which I would like to mostly avoid, except for the `Any` interface with a specific type interface.

`Value` suffix actually causes issues in my own project, since I use https://pkg.go.dev/github.com/jackc/pgx/v5@v5.5.0/pgtype 

What about an `As` prefix? `AsAny()`, `AsInt()`? Or something else entirely?

Reply:
Good point. 

Yes `As` looks like a good short alternative.

Reply:
a7e4934bf2f579623ebe06f49c2139a716e3697e